aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/model/transform/XmlResponsesSaxParser.java

```diff
@@ -157,63 +159,73 @@ protected void parseXmlInputStream(DefaultHandler handler, InputStream inputStre
         }
     }
 
-    protected InputStream sanitizeXmlDocument(DefaultHandler handler, InputStream inputStream)
-            throws IOException {
+    protected StringReader sanitizeXmlDocument(DefaultHandler handler, InputStream inputStream)
+        throws IOException {
+        if (log.isDebugEnabled()) {
+            log.debug("Sanitizing XML document destined for handler " + handler.getClass());
+        }
+        try {
 
-        if (!sanitizeXmlDocument) {
-            // No sanitizing will be performed, return the original input stream unchanged.
-            return inputStream;
-        } else {
-            if (log.isDebugEnabled()) {
-                log.debug("Sanitizing XML document destined for handler " + handler.getClass());
+             /*
+              * Read object listing XML document from input stream provided into a
+              * string buffer, so we can replace troublesome characters before
+              * sending the document to the XML parser.
+              */
+            StringBuilder listingDocBuffer = new StringBuilder();
+            BufferedReader br = new BufferedReader(
+                new InputStreamReader(inputStream, Constants.DEFAULT_ENCODING));
+
+            char[] buf = new char[8192];
+            int read = -1;
+            while ((read = br.read(buf)) != -1) {
+                listingDocBuffer.append(buf, 0, read);
+            }
+            br.close();
```

Comment:
Thank you for the review @steveloughran and @liuml07 . Addressed the review comments in the latest commit.